src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                    }
+
+                    // Notify the other threads waiting for the initialization at start().
+                    this.owner.startTimeInitialized.Signal();
+
+                    do
+                    {
+                        TimeSpan deadline = this.WaitForNextTick();
+                        if (deadline > TimeSpan.Zero)
+                        {
+                            int idx = (int)(this.tick & this.owner.mask);
+                            this.ProcessCancelledTasks();
+                            HashedWheelBucket bucket = this.owner.wheel[idx];
+                            this.TransferTimeoutsToBuckets();
+                            bucket.ExpireTimeouts(deadline);
+                            this.tick++;
+                        }
+                    }
+                    while (Volatile.Read(ref this.owner.workerStateVolatile) == WorkerStateStarted);
+
+                    // Fill the unprocessedTimeouts so we can return them from stop() method.
+                    foreach (HashedWheelBucket bucket in this.owner.wheel)
+                    {
+                        bucket.ClearTimeouts(this.unprocessedTimeouts);
+                    }
+                    while (true)
+                    {
+                        HashedWheelTimeout timeout;
+                        if (!this.owner.timeouts.TryDequeue(out timeout))
+                        {
+                            break;
+                        }
+                        if (!timeout.Cancelled)
+                        {
+                            this.unprocessedTimeouts.Add(timeout);
+                        }
+                    }
+                    this.ProcessCancelledTasks();
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error("Timeout processing failed.", ex);
+                }
+            }
+
+            void TransferTimeoutsToBuckets()
+            {
+                // transfer only max. 100000 timeouts per tick to prevent a thread to stale the workerThread when it just
+                // adds new timeouts in a loop.
+                for (int i = 0; i < 100000; i++)
+                {
+                    HashedWheelTimeout timeout;
+                    if (!this.owner.timeouts.TryDequeue(out timeout))
+                    {
+                        // all processed
+                        break;
+                    }
+                    if (timeout.State == HashedWheelTimeout.StCancelled)
+                    {
+                        // Was cancelled in the meantime.
+                        continue;
+                    }
+
+                    long calculated = (timeout.Deadline.Ticks + this.owner.tickDuration - 1) / this.owner.tickDuration; // ceiling to timeout later rather than earlier
+                    timeout.RemainingRounds = (calculated - this.tick) / this.owner.wheel.Length;
```

Comment:
yes and it's supported

---
In reply to: [114476429](https://github.com/Azure/DotNetty/pull/242#discussion_r114476429) [](ancestors = 114476429)